Realm/Realm/Realm.cs

```diff
@@ -1357,6 +1344,54 @@ internal void ExecuteOutsideTransaction(Action action)
 
         #endregion Transactions
 
+        internal class SessionProvider
+        {
+            private WeakReference<Session>? _weakSessionRef;
+            private Session? _strongSessionRef;
+
+            public Session GetSession(SharedRealmHandle handle)
+            {
+                if(_strongSessionRef?.IsClosed == false)
+                {
+                    return _strongSessionRef;
+                }
+                else if (_weakSessionRef?.TryGetTarget(out var targetSession) == true && !targetSession.IsClosed)
+                {
+                    return targetSession;
+                }
+
+                var session = new Session(handle.GetSession());
+                session.Subscribed += Session_Subscribed;
+                session.Unsubscribed += Session_Unsubscribed;
+
+                if (_weakSessionRef is null)
+                {
+                    _weakSessionRef = new WeakReference<Session>(session);
+                }
+                else
+                {
+                    _weakSessionRef.SetTarget(session);
+                }
```

Comment:
Why not allocate a new reference in both cases? Are we winning a lot by calling `SetTarget`?

Reply:
I took the code the code that was already there, and I thought it made sense, so we avoid having the GC collecting all those "old" `WeakReference`?

Reply:
I guess whoever wrote the old code must have been very smart 🤓 Perhaps a bit too smart - I'd be surprised if that ever made a difference as it'd only trigger if the user is accessing the session property multiple times and letting it get GC-ed in between, so clearly not a hot path. In any case, since it's moved code, you don't need to simplify it unless you want to.

Reply:
Yeah, I agree, premature optimisation probably, I've removed it 😁